### ISSUE TITLE
Prompt user to run "brew upgrade" after brew update is complete

### ISSUE THREAD

**Issue by rwuwon** [Homebrew/brew#5400]
# A detailed description of the proposed feature

Improve the user interface so that after running `brew update`, the user knows that there are upgradeable packages. Let the new user know that a `brew upgrade` command is probably immediately required/desired.

# The motivation for the feature

Many other package managers (apt, dnf, pacman -Syu) make this clear from the outset. Unfortunately, while the Brew [homepage](https://brew.sh/) looks very friendly and inviting, this can easily lead a new user to a false sense of security and not realise that `brew update` and `brew upgrade` should be a part of normal usage. The user may also require all the reminders they can get that Brew packages update separately to the rest of macOS and that `brew update` alone is not enough.

If a user runs `brew update` twice in succession without running `brew upgrade`, I recall the output becomes `Already up-to-date.` - which might give the impression that `brew upgrade` is not needed.

# How the feature would be relevant to at least 90% of Homebrew users

I estimate 90% or more of new Homebrew users would be unaware that `brew upgrade` is required if they have been familiar with all-in-one commands such as `dnf update`, `pacman -Syu` or `apt update && apt upgrade` commands.

As an example, running `apt update` on a Debian system may result in the following type of line: `6 packages can be upgraded. Run 'apt list --upgradable' to see them.` Essentially, this is all that is needed to prompt the user into doing more.

# What alternatives to the feature have been considered

Update the documentation to make `brew upgrade` a more essential command. It is currently not listed as "essential" in the [manual](https://docs.brew.sh/Manpage) - in fact, it is buried halfway down that page. Interestingly however, `brew --help` *does* list it near the top in the *Example usage*, so there is a discrepancy here between the --help and the man/homepage.

In summary, the current design could contain more [discoverability](https://fishshell.com/docs/current/design.html#disc) and for a utility of this type that caters towards ease-of-use (as evidenced by the friendliness of https://brew.sh and other tutorials that might make passing mention of Homebrew and assume existing knowledge), it should be reasonable to expect to not need to find out that `brew upgrade` is a command that one can only realise after reading through an entire manual.

**Comment by vitorgalvao** [comment]
> so that after running `brew update`, the user knows that there are upgradeable packages.

That already happens. If any installed formulae have an update, they are listed in bold.

> `brew update` and `brew upgrade` should be a part of normal usage.

Not really. `brew update` is (by default) ran automatically. In normal usage, you can just run `brew upgrade` and not worry about ever running `update`.

> The user may also require all the reminders they can get that Brew packages update separately to the rest of macOS

I don’t have number, but I doubt there are many users that deliberately run a terminal command to install an external tool and then think it auto updates together with macOS. Even regular macOS apps need to prompt you to update.

**Comment by chdiza** [comment]
Running `upgrade` is dangerous.  Nothing at all should be added.

Time and time again, some upstream makes a breaking change that e.g. irreversibly migrates somebody's database to an undesired format or something.  But people blindly run `upgrade` and then complain *here* even though they didn't bother to check the release notes to see what's new in the new version.  Facepalm.

Don't add to this by *advising* them to just hurry up and `upgrade`.

**Comment by rwuwon** [comment]
@chdiza that is a surprise to me given that the manual and quick help (brew -h) makes no warning of the sort in the section for upgrade.

Would a more descriptive blurb that MikeMcQuaid suggested be problematic?

For example:
"You have X outdated kegs: x y z

Please review before running brew upgrade".

**Comment by vitorgalvao** [comment]
> the manual and quick help (brew -h) makes no warning of the sort

Because that’s not really a problem/limitation with Homebrew. It’s something inherent to (and out of the control of) package managers (mostly). Having a “be careful with what you upgrade” warning would likely create more problems (fear of upgrading) than solving.

Is this really a common problem?

It is not uncommon for a particular user to be faced with something they didn’t understand from the documentation alone, eventually figure it out (kudos!), and then ask for/suggest such clarification in the documentation. I applaud the spirit — they tend to be respectful and well-intentioned users with a positive tone and a thorough explanation of the problem they’ve encountered (all boxes you’ve ticked!). But they also tend to be isolated cases that eventually figured out the solution on their own. Those traits together pretty much invalidate the need for the requested documentation. And speaking for the Homebrew Cask side (my area), if we always added the documentation in these cases, we’d be pretty much explaining basic ruby and bash features (not an exaggeration). We have to draw the line *somewhere*.

Adding the extra explanatory line in `update` might indeed help, but maybe it should happen if we verify this to be a more common source of confusion?

**Comment by MikeMcQuaid** [comment]
The one-liner is going to be more involved (and not consistently correct and/or slow down update for everyone) so: passing on this suggestion.
